swayfmt/src/comments.rs

```diff
@@ -0,0 +1,66 @@
+use std::fmt::Write;
+use sway_ast::token::Comment;
+use sway_types::Spanned;
+
+use crate::{formatter::FormattedCode, Formatter, FormatterError};
+
+// Given a start, an end and a CommentMap, return references to all comments
+// contained within a given start and end of a span, in an exclusive range.
+pub fn get_comments_between(start: usize, end: usize, formatter: &mut Formatter) -> Vec<&Comment> {
+    let mut comments = vec![];
+    let iter = formatter.comment_map.clone().into_keys();
+
+    for bs in iter {
+        if bs.start > start && bs.end < end {
```

Comment:
The `comment_map` is a `BTreeMap`. The `BTreeMap` stores its elements using ordered keys, and actually [provides a dedicated method](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) for efficiently retrieving elements within a given range.

Reply:
So to be able to get a range over `CommentMap` I had to implement `RangeBounds` for `ByteSpan`, but this means the range over the `CommentMap` takes a starting `ByteSpan` and an ending `ByteSpan`. We not only want the comments between `ByteSpans` but we also might want comments within a `ByteSpan`. Would this still be appropriate here? eg. within the `IfExpr` i'd like to check if there are comments self-contained between `self.span().start` and `self.span().end`.

Reply:
Ahh good point! For some reason I was thinking of the keys as indices rather than the full byte spans. In that case, perhaps we are best to keep this `comments_between` function (though have it take a `std::ops::Range<usize>` as input and return an iterator).